### PR TITLE
optimize: make longest_road faster - was previously O(n^2) or worse

### PR DESCRIPTION
**Pull request by Luke-Nukem**
It was iterating through the grid once for each player, then running the recursive algorithm from both start and end of unboken road. This was honestly a pretty horrible hack.

New code iterates through the grid once only, and runs the recursive algorithm *once* only for each *start* of a road, be it a single road or a chain. However if the road is something like a 'Y' shape it may run twice, but for straight roads (or upside down 'Y') it is much more efficient.

Does need a bit more testing to make sure there are no regressions, if there are the problems will start on line 262 of board.js

### CODE REVIEW

File: app/game/state_machine.js [hunk 1]
```diff
@@ -769,10 +771,11 @@ StateMachine.prototype.buy_dev_card = function (data){
         data_package.player = this.game.players[data.player_id];
 
         //  Refreshes all player's scores, strip out to calc only one players score :- TODO
+        var longest_road_map = this.game.board.longest_roads(this.game.players);
         for (var p=0; p< this.game.players.length; p++) {
           var player_id = this.game.players[p].id;
-          this.game.players[p].score.longest_road = this.game.board.longest_road_for_player(player_id);
-        };
+          this.game.players[p].score.longest_road = longest_road_map.get(player_id);
+        }
```
Comment (sumnerfit) [inline app/game/state_machine.js]: Is the section from line 774 -778 identical to 257 -261?
If so, can it be pulled out to it's own function?

Comment (Luke-Nukem) [reply app/game/state_machine.js]: Yeah doing a fair clean up there.

**Comment by Luke-Nukem** [comment]
Also appears that it isn't counting just the longest path, but all path that are connected - eg, if a road is shaped like a 'Y' with 9 sections in total, it will count 9, not just the longest path.

**Comment by timack** [comment]
Taking a look at this now.

**Comment by timack** [comment]
@Luke-Nukem Doesn't seem to be working. I created a road of 6 segments and I wasn't awarded the longest road.

image

**Comment by Luke-Nukem** [comment]
Oh... It definitely scores right. I don't know much about the front-end
code. Can we get this merged and someone else can take a quick look at that?

**Comment by Luke-Nukem** [comment]
@timack that should be fixed now - silly me had removed the field in the player struct that the scoring depended on. Actual scoring was working fine, but the bit that the GUI relied on was removed.

**Comment by timack** [comment]
@Luke-Nukem Still hitting some issues ...

This happens when I try and end turn on the player that hasn't placed any roads this turn

```/Users/tim/Documents/Personal/Massey/capstone/public/data_api/board.js:304
  for (var r=0; r<node.n_roads.length; r++) {
                              ^

TypeError: Cannot read property 'length' of undefined
    at Board.longest_road_for_player (/Users/tim/Documents/Personal/Massey/capstone/public/data_api/board.js:304:31)
    at StateMachine.tick (/Users/tim/Documents/Personal/Massey/capstone/app/game/state_machine.js:271:71)
    at Socket.<anonymous> (/Users/tim/Documents/Personal/Massey/capstone/app/game/games.js:108:23)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
    at /Users/tim/Documents/Personal/Massey/capstone/node_modules/socket.io/lib/socket.js:514:12
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! capstone@0.0.1 start: `node app/app.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the capstone@0.0.1 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/tim/.npm/_logs/2017-09-23T02_18_43_087Z-debug.log```

**Comment by Luke-Nukem** [comment]
Last issue was caused by misuse of one of the longest road functions in another commit. Now fixed.

**Pull request merged**